### ISSUE TITLE
Intermittent timed out accessing nodelocaldns

### ISSUE THREAD

**Issue by ausias-armesto** [kubernetes/dns#598]
Hi,

I'm facing a reccurrent timeout issue when **Github runners** pods try to resolve DNS from nodelocaldns. It is not happenning always but very often and make the pods to run into serious issues.

Steps to reproduce the problem

- Connect to the Github runner shell
- Execute the following command `telnet 10.1.0.10 53`
- Watch the nodelocaldns logs for that server.
- I notice that the first log entry does not appear until some seconds later
- Sometimes the logs don't show anything on nodelocaldns because the telnet request timed out.

Infrastructure Setup:

- Kubernetes nodes on Debian 11 Bulseye
- Kubernetes version 1.24.6
- CNI: Calico v3.23.3
- Distribution: Kubespray on bare metal

<details>
  <summary>This is my Corefile configuration </summary>
 
```
    cluster.prod:53 {
        errors
        cache {
            success 9984 30
            denial 9984 5
        }
        reload
        loop
        bind 10.1.0.10
        forward . 10.1.0.3 {
            force_tcp
        }
        prometheus :9253
        #health 10.1.0.10:9254
    }
    in-addr.arpa:53 {
        errors
        cache 30
        reload
        loop
        bind 10.1.0.10
        forward . 10.1.0.3 {
            force_tcp
        }
        prometheus :9253
    }
    ip6.arpa:53 {
        errors
        cache 30
        reload
        loop
        bind 10.1.0.10
        forward . 10.1.0.3 {
            force_tcp
        }
        prometheus :9253
    }
    .:53 {
        errors
        cache 30
        reload
        loop
        log
        bind 10.1.0.10
        forward . 8.8.8.8 8.8.4.4
        prometheus :9253
    }

```
</details>

The Kubelet-configuration has this attribute
```
clusterDNS:
- 10.1.0.10
```

Virtual interface details:
```
root@server-prod-05 /etc/kubernetes # ip addr | grep nodelocaldns -A 3
142: nodelocaldns: <BROADCAST,NOARP> mtu 1500 qdisc noqueue state DOWN group default 
    link/ether 52:9b:2e:fa:6b:31 brd ff:ff:ff:ff:ff:ff
    inet 10.1.0.10/32 scope global nodelocaldns
       valid_lft forever preferred_lft forever

```

<details>
  <summary>This is the pod definition:</summary>
  

	```
	apiVersion: v1
	kind: Pod
	metadata:
	  annotations:
	    actions-runner-controller/token-expires-at: "2023-09-07T09:45:09+02:00"
	    actions-runner/id: "15941"
	    cni.projectcalico.org/containerID: c14c3aa60c8ccb16b62d32f856b67797faa89ef1a173fd942c6374438a0dddea
	    cni.projectcalico.org/podIP: 10.2.154.20/32
	    cni.projectcalico.org/podIPs: 10.2.154.20/32
	    sync-time: "2023-09-07T06:50:19Z"
	  creationTimestamp: "2023-09-07T06:50:23Z"
	  finalizers:
	  - actions.summerwind.dev/runner-pod
	  labels:
	    actions-runner: ""
	    actions-runner-controller/inject-registration-token: "true"
	    pod-template-hash: 76657dc54
	    runner-deployment-name: self-hosted-ubuntu-2cpu
	    runner-template-hash: 6f9c9448c9
	  name: self-hosted-ubuntu-2cpu-9sn6h-9zm75
	  namespace: github-runners
	  ownerReferences:
	  - apiVersion: actions.summerwind.dev/v1alpha1
	    blockOwnerDeletion: true
	    controller: true
	    kind: Runner
	    name: self-hosted-ubuntu-2cpu-9sn6h-9zm75
	    uid: 523d0e3d-23d5-42fa-b48f-e43eef8c0a07
	  resourceVersion: "99700756"
	  uid: 8140f353-cdf9-4f9b-85db-07f41b99c9ad
	spec:
	  containers:
	  - env:
	    - name: RUNNER_ORG
	      value: hoprnet
	    - name: RUNNER_REPO
	    - name: RUNNER_ENTERPRISE
	    - name: RUNNER_LABELS
	      value: ubuntu-2-core
	    - name: RUNNER_GROUP
	      value: hoprnet
	    - name: DOCKER_ENABLED
	      value: "true"
	    - name: DOCKERD_IN_RUNNER
	      value: "false"
	    - name: GITHUB_URL
	      value: https://github.com/
	    - name: RUNNER_WORKDIR
	      value: /runner/_work
	    - name: RUNNER_EPHEMERAL
	      value: "true"
	    - name: RUNNER_STATUS_UPDATE_HOOK
	      value: "false"
	    - name: GITHUB_ACTIONS_RUNNER_EXTRA_USER_AGENT
	      value: actions-runner-controller/v0.27.4
	    - name: DOCKER_HOST
	      value: unix:///run/docker.sock
	    - name: RUNNER_NAME
	      value: self-hosted-ubuntu-2cpu-9sn6h-9zm75
	    - name: RUNNER_TOKEN
	      value: BAXWHNTFM7ILUZR5ZNDREVDE7F7YK
	    image: europe-west3-docker.pkg.dev/hoprassociation/docker-images/gh-runner:latest
	    imagePullPolicy: Always
	    name: runner
	    resources:
	      limits:
	        cpu: "2"
	        memory: 4Gi
	      requests:
	        cpu: "2"
	        memory: 2Gi
	    securityContext: {}
	    terminationMessagePath: /dev/termination-log
	    terminationMessagePolicy: File
	    volumeMounts:
	    - mountPath: /runner
	      name: runner
	    - mountPath: /runner/_work
	      name: work
	    - mountPath: /run
	      name: var-run
	    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
	      name: kube-api-access-qr58s
	      readOnly: true
	  - args:
	    - dockerd
	    - --host=unix:///run/docker.sock
	    - --group=$(DOCKER_GROUP_GID)
	    - --mtu
	    - "1400"
	    env:
	    - name: DOCKER_GROUP_GID
	      value: "1001"
	    - name: DOCKERD_ROOTLESS_ROOTLESSKIT_MTU
	      value: "1400"
	    image: docker:dind
	    imagePullPolicy: IfNotPresent
	    lifecycle:
	      preStop:
	        exec:
	          command:
	          - /bin/sh
	          - -c
	          - timeout "${RUNNER_GRACEFUL_STOP_TIMEOUT:-15}" /bin/sh -c "echo 'Prestop
	            hook started'; while [ -f /runner/.runner ]; do sleep 1; done; echo 'Waiting
	            for dockerd to start'; while ! pgrep -x dockerd; do sleep 1; done; echo
	            'Prestop hook stopped'" >/proc/1/fd/1 2>&1
	    name: docker
	    resources: {}
	    securityContext:
	      privileged: true
	    terminationMessagePath: /dev/termination-log
	    terminationMessagePolicy: File
	    volumeMounts:
	    - mountPath: /runner
	      name: runner
	    - mountPath: /run
	      name: var-run
	    - mountPath: /runner/_work
	      name: work
	    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
	      name: kube-api-access-qr58s
	      readOnly: true
	  dnsPolicy: ClusterFirst
	  enableServiceLinks: true
	  nodeName: server-prod-05
	  preemptionPolicy: PreemptLowerPriority
	  priority: 0
	  restartPolicy: Never
	  schedulerName: default-scheduler
	  securityContext: {}
	  serviceAccount: github-runners-sa
	  serviceAccountName: github-runners-sa
	  terminationGracePeriodSeconds: 30
	  tolerations:
	  - effect: NoExecute
	    key: node.kubernetes.io/not-ready
	    operator: Exists
	    tolerationSeconds: 300
	  - effect: NoExecute
	    key: node.kubernetes.io/unreachable
	    operator: Exists
	    tolerationSeconds: 300
	  volumes:
	  - emptyDir: {}
	    name: runner
	  - emptyDir: {}
	    name: work
	  - emptyDir:
	      medium: Memory
	      sizeLimit: 1M
	    name: var-run
	  - name: kube-api-access-qr58s
	    projected:
	      defaultMode: 420
	      sources:
	      - serviceAccountToken:
	          expirationSeconds: 3607
	          path: token
	      - configMap:
	          items:
	          - key: ca.crt
	            path: ca.crt
	          name: kube-root-ca.crt
	      - downwardAPI:
	          items:
	          - fieldRef:
	              apiVersion: v1
	              fieldPath: metadata.namespace
	            path: namespace
	```

</details>

<details>
  <summary>This is configuration for the Custom Dockerfile</summary>
 
```
FROM summerwind/actions-runner:ubuntu-20.04

RUN \
    # Install github cli using brew
    sudo /bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)" \
    && eval "$(/home/linuxbrew/.linuxbrew/bin/brew shellenv)" \
    && sudo brew install gh \
    && sudo ln -s /home/linuxbrew/.linuxbrew/bin/gh /usr/bin/gh \
    ## Install yq command
    && wget https://github.com/mikefarah/yq/releases/download/v4.34.2/yq_linux_amd64.tar.gz -O /tmp/yq_linux_amd64.tar.gz \
    && tar xzf /tmp/yq_linux_amd64.tar.gz --directory /tmp \
    && sudo mv /tmp/yq_linux_amd64 /usr/bin/yq \
    # Install packages
    && sudo apt-get update \
    && sudo apt install -y bc bsdmainutils lsof \
    # Install Kubectl
    && sudo curl --silent -o /usr/bin/kubectl -LO "https://dl.k8s.io/release/$(curl --silent -L -s https://dl.k8s.io/release/stable.txt)/bin/linux/amd64/kubectl" \
    && sudo chmod +x /usr/bin/kubectl \
    # Clean up
    && sudo rm -rf /var/lib/apt/lists/* \
    && sudo apt-get purge -y --auto-remove -o APT::AutoRemove::RecommendsImportant=false

```
</details>

**Additional tests**
- I've changed the MTU at the interface to be align with the MTU of the pod but didn't work, I also modified the MTU at the pod to be align with the one used at the interface and didn't work either.
- I've changed the pod DNSPolicy to all posible values, but didn't work.
- I have disabled the firewall, but it was not either the cause
- I've tested the `telnet 10.1.0.10 53` from a shell directly connecting to the worker node, and it works correctly
- I've tested the `telnet 10.1.0.10 53` from other pod running in the same node, and it works correctly
- The error happens in all nodes, not in a specific one
- I've analysed traffic using `tcpdump` from the worker node, and doesn't appear any traffic to destination 10.1.0.10
- The grafana monitoring metrics for dropped packets is 0
- I've executed the `tcpdump dst 10.1.0.10` within the docker container, and with the same docker container but different shell I've performed the `telnet 10.1.0.10 53`. Then the tcpdump didn't start to bump data after some seconds (5-10 seconds), time spent the telnet command until opens the connection
- I've change to the default docker image `summerwind/actions-runner:ubuntu-20.04` and the problem is the same

**Comment by BenB196** [comment]
I'm noticing a similar issue here when working with GitLab Runners. Was curious if there has been any movement on this one. I notice it on both self-host Kubernetes cluster (RKE2) and AWS EKS clusters.

**Comment by ausias-armesto** [comment]
I managed to solve it.
I identified initially the problem as related with localdns but the root cause was an incompatible issue with DockerInDocker container of the self hosted runner. If the container is deployed as a side-car container, then the networking at the pod was not working correctly.
The initial issue was identified as a problem with nodelocaldns but it was not working many other things like:

- Ping to external nodes. Executing this command `ping 8.8.8.8 -c 5` resulted in a timeout and host not reachable.
- Telnet to nodelocaldns `telnet 10.1.0.10 53`
- Telnet to coredns `telnet 10.1.0.3 53`
- Ping to other pod running in the same node: `ping 10.2.0.134`
- Ping from other pod running in the same node to the github runner didn't work either

So, in order to remove the side-container it was needed to set this attribute:
```
dockerEnabled: false
```
And then in the Github action workflow, specify to use the Kubernetes driver for the docker Build X action.
```

      - name: Set up Docker Buildx
        uses: docker/setup-buildx-action@v2
        with:
          driver: kubernetes
```
I suppose that this step will be different in Gitlab.

Finally, check the MTU from the container and the Kubernetes cluster because I had to customised it with this attribute:
```
dockerMTU: 1400
```
